### ISSUE TITLE
Robustness issue in QuadPrefixTree

### ISSUE THREAD

**Issue by chenhh021** [apache/lucenenet#644]
Hi,

when trying to index a point with coordinate (144502.06, 639062.07) in Lucene.Net.Spatial, a System.ArgumentOutOfRangeException is thrown out.  The stack trace is listed below. This bug can be reproduced by [this test](https://github.com/apache/lucenenet/commit/956c78c9eed7548a1069f5633510c7aa351a9ad4).

`System.ArgumentOutOfRangeException: Index was out of range. Must be non-negative and less than the size of the collection. (Parameter 'index')
Actual value was 0.
   at J2N.Collections.Generic.List‘1.get_Item(Int32 index)
   at Lucene.Net.Spatial.Prefix.Tree.QuadPrefixTree.GetCell(IPoint p, Int32 level) in D:\chh\SE\projects\lucenenet\src\Lucene.Net.Spatial\Prefix\Tree\QuadPrefixTree.cs:line 167
   at Lucene.Net.Spatial.Prefix.Tree.SpatialPrefixTree.GetCells(IPoint p, Int32 detailLevel, Boolean inclParents) in D:\chh\SE\projects\lucenenet\src\Lucene.Net.Spatial\Prefix\Tree\SpatialPrefixTree.cs:line 286
   at Lucene.Net.Spatial.Prefix.Tree.SpatialPrefixTree.GetCells(IShape shape, Int32 detailLevel, Boolean inclParents, Boolean simplify) in D:\chh\SE\projects\lucenenet\src\Lucene.Net.Spatial\Prefix\Tree\SpatialPrefixTree.cs:line 204
   at Lucene.Net.Spatial.Prefix.PrefixTreeStrategy.CreateIndexableFields(IShape shape, Double distErr) in D:\chh\SE\projects\lucenenet\src\Lucene.Net.Spatial\Prefix\PrefixTreeStrategy.cs:line 139
   at Lucene.Net.Spatial.Prefix.PrefixTreeStrategy.CreateIndexableFields(IShape shape) in D:\chh\SE\projects\lucenenet\src\Lucene.Net.Spatial\Prefix\PrefixTreeStrategy.cs:line 133
   at Lucene.Net.Spatial.StrategyTestCase.newDoc(String id, IShape shape) in D:\chh\SE\projects\lucenenet\src\Lucene.Net.Tests.Spatial\StrategyTestCase.cs:line 220
   at Lucene.Net.Spatial.StrategyTestCase.adoc(String id, IShape shape) in D:\chh\SE\projects\lucenenet\src\Lucene.Net.Tests.Spatial\StrategyTestCase.cs:line 211
   at Lucene.Net.Spatial.RobustnessTest.QuadTreeRobustness() in D:\chh\SE\projects\lucenenet\src\Lucene.Net.Tests.Spatial\RobustnessTest.cs:line 46`

Configurations:
```c#
SpatialContextFactory test_factory = new SpatialContextFactory();
test_factory.IsGeo = false;
test_factory.WorldBounds = new Rectangle(0.0, 700000.0, 0.0, 650000.0, null);

SpatialContext ctx = new SpatialContext(test_factory);
SpatialPrefixTree grid; SpatialStrategy strategy;

grid = new QuadPrefixTree(ctx, 40);
strategy = new RecursivePrefixTreeStrategy(grid, "recursive_quad");
```

I am new to the community, please let me know if I did anything improper. Thanks in advance!

**Comment by chenhh021** [comment]
@NightOwl888 @rclabo could you please have a look at this problem?

**Comment by chenhh021** [comment]
I find lucene ever fix the same bug via this [commit](https://github.com/apache/lucene/commit/26628b2717a73235db56fde94f7f5b64cbc5b8b2) . Although it was done for lucene 9.4, the changed code is shared by the present version of lucene.net. 

I have created a [PR](https://github.com/apache/lucenenet/pull/738) mimic that commit. What I have done include:

1.Reimplement QuadPrefixTree's getCell method;
2. Add a version check so the older versions still use the older implementation.

Could anyone take a review?

**Comment by NightOwl888** [comment]
This is a duplicate of [LUCENE-8755(https://issues.apache.org/jira/browse/LUCENE-8755), and will be fixed when we upgrade to 8.3.0. For now, we will be sticking with the 4.8.0 behavior.

For those who wish to have a patch before then, the `AbstractPrefixTreeFilter` and `SpatialPrefixTree` are both abstract and the `QuadPrefixTree` is not sealed and all members are virtual so it can be extended to include the patch, if necessary.
